compiler/rustc_ast/src/util/comments.rs

```diff
@@ -24,6 +24,14 @@ pub struct Comment {
     pub pos: BytePos,
 }
 
+/// A fast conservative estimate on whether the string can contain documentation links.
+/// Doc links are markdown links (<https://spec.commonmark.org/0.30/#links>) so either
+/// a `[]` pair or `<>` pair must exist in the string.
+#[inline]
+pub fn may_have_doc_links(s: &str) -> bool {
+    s.contains(&['[', '<']) && s.contains(&[']', '>'])
```

Comment:
`<>` links aren't being used for intra doc links but only "as is".

Reply:
Apparently they still go through path resolution?
Because if they are filtered away then tests like `src/test/rustdoc-ui/intra-doc/email-address-localhost.rs` fail.
See https://github.com/rust-lang/rust/pull/83865#issuecomment-1100424663.

If we can change the behavior in cases like `email-address-localhost.rs` and skip comments with `<>`s but without `[]`s entirely that would be really great, because `<>`s are almost always false positives.

Reply:
What the hell... Do you want me to send a PR to remove `<>` check in intra-doc links? It wouldn't change much in here but it'd be still that less to worry about...

Reply:
Yes, it would be ideal if links without `[]` never showed up in the output from `fn preprocess_link` or even `fn markdown_links`.

Reply:
What? No, it's wrong and inconsistent to treat `[]` different from `<>`. Please don't break stable behavior.

Reply:
On stable `<>` links don't show in the visible output tho.
It means they are removed from the output by some other pass.

Reply:
Ok, I think I see what's going on. Pulldown-cmark only passes autolinks to rustdoc if it thinks they're already links, our `broken_link_callback` doesn't seem to be called. There's lots of existing buggy behavior around this, e.g.
```rust
/// <fn@localhost>
pub fn localhost() {}
```
generates a link to `mailto:fn.localhost.html` which is definitely wrong. We should investigate and fix those bugs before changing the behavior further.

Reply:
It's worth a discussion apparently. :)

> On stable `<>` links don't show up in the visible output tho.
> It means they are removed from the output by some other pass.

Just to confirm @petrochenkov's words:

```
//! This is a super <https://crate.com> <Foo>.

pub struct Foo;
```

Gives:

![Screenshot from 2022-04-18 16-07-21](https://user-images.githubusercontent.com/3050060/163820078-e1c513dc-851f-4045-92cf-88e3280f00ee.png)

@jyn514: Also, unlinke `[]`, `<>` can only be HTML links and we don't mention anywhere that they can be used for intra-doc links (based on https://doc.rust-lang.org/rustdoc/linking-to-items-by-name.html).

But you're right on fixing the other bugs first before deciding about this. What do you think should be done next?

Reply:
I don't want to be responsible for this decision.

Reply:
It's not about being responsible but having your opinion on this topic. You are a very important rustdoc contributor and your opinion is highly respected. Hence why I'm asking you. You said "don't break stable behaviour" but it's already not supported for `<>` to be intra-doc links. So at this point, I think it'd be only logical to make it official, which would allow us some potential perf improvements. But maybe I missed something and it's why I want to hear your opinion. Then I can bring it up to the rest of the rustdoc team with all elements in hand.

Reply:
I opened this [zulip stream](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/Only.20support.20intra-doc.20links.20for.20.60.5B.5D.60) to talk about it.

Reply:
We should officially not support intra-doc links in `<>`. It’s not documented, it’s buggy, and the `<>` symbols are already painfully overloaded. What's supposed to happen when someone tries to use a `<b>` HTML tag, but they have a function named `b()` in scope?

Just to get across how confusing it is, *you both got it wrong.* `<Foo>` is not being parsed as an intra-doc link. It's an HTML tag. It's not getting stripped. It's right there in the generated HTML:

```html
<p>This is a super <a href="https://crate.com">https://crate.com</a> <Foo>.</p>
```

Reply:
More importantly, the RFC says that auto-links shouldn't work: https://rust-lang.github.io/rfcs/1946-intra-rustdoc-links.html#no-autolinks-style

Reply:
If the RFC says it, then I'll send a PR to clean things up a bit.

Reply:
Changed the condition to only consider square brackets.